packages/gatsby-core-utils/src/remote-file-utils/fetch-file.ts

```diff
@@ -0,0 +1,269 @@
+import fs from "fs-extra"
+import type { IncomingMessage } from "http"
+import type { Headers, Options } from "got"
+import type { GatsbyCache } from "gatsby"
+
+// keeping the I for backward compatibility
+export type IFetchRemoteFileOptions = {
+  url: string
+  auth?: {
+    htaccess_pass?: string
+    htaccess_user?: string
+  }
+  httpHeaders?: Headers
+  ext?: string
+  name?: string
+  cacheKey?: string
+} & (
+  | {
+      directory: string
+      cache?: never
+    }
+  | {
+      cache?: GatsbyCache
+      directory: never
+    }
+)
+
+const STALL_RETRY_LIMIT = process.env.GATSBY_STALL_RETRY_LIMIT
+  ? parseInt(process.env.GATSBY_STALL_RETRY_LIMIT, 10)
+  : 3
+const STALL_TIMEOUT = process.env.GATSBY_STALL_TIMEOUT
+  ? parseInt(process.env.GATSBY_STALL_TIMEOUT, 10)
+  : 30000
+
+const CONNECTION_TIMEOUT = process.env.GATSBY_CONNECTION_TIMEOUT
+  ? parseInt(process.env.GATSBY_CONNECTION_TIMEOUT, 10)
+  : 30000
+
+const INCOMPLETE_RETRY_LIMIT = process.env.GATSBY_INCOMPLETE_RETRY_LIMIT
+  ? parseInt(process.env.GATSBY_INCOMPLETE_RETRY_LIMIT, 10)
+  : 3
+
+// jest doesn't allow us to run all timings infinitely, so we set it 0  in tests
+const BACKOFF_TIME = process.env.NODE_ENV === `test` ? 0 : 1000
+
+function range(start: number, end: number): Array<number> {
+  return Array(end - start)
+    .fill(null)
+    .map((_, i) => start + i)
+}
+
+// Based on the defaults of https://github.com/JustinBeckwith/retry-axios
+const STATUS_CODES_TO_RETRY = [...range(100, 200), 429, ...range(500, 600)]
+const ERROR_CODES_TO_RETRY = [
+  `ETIMEDOUT`,
+  `ECONNRESET`,
+  `EADDRINUSE`,
+  `ECONNREFUSED`,
+  `EPIPE`,
+  `ENOTFOUND`,
+  `ENETUNREACH`,
+  `EAI_AGAIN`,
+  `ERR_NON_2XX_3XX_RESPONSE`,
+  `ERR_GOT_REQUEST_ERROR`,
+]
+
+/**
+ * requestRemoteNode
+ * --
+ * Download the requested file
+ *
+ * @param  {String}   url
+ * @param  {Headers}  headers
+ * @param  {String}   tmpFilename
+ * @param  {Object}   httpOptions
+ * @param  {number}   attempt
+ * @return {Promise<Object>}  Resolves with the [http Result Object]{@link https://nodejs.org/api/http.html#http_class_http_serverresponse}
+ */
+export async function requestRemoteNode(
+  url: string | URL,
+  headers: Headers,
+  tmpFilename: string,
+  httpOptions?: Options,
+  attempt: number = 1
+): Promise<IncomingMessage> {
+  const { default: got, RequestError } = await import(`got`)
+
+  return new Promise((resolve, reject) => {
+    let timeout: NodeJS.Timeout
+    const fsWriteStream = fs.createWriteStream(tmpFilename)
+    fsWriteStream.on(`error`, (error: unknown) => {
+      if (timeout) {
+        clearTimeout(timeout)
+      }
+
+      reject(error)
+    })
+
+    // Called if we stall for 30s without receiving any data
+    const handleTimeout = async (): Promise<void> => {
+      fsWriteStream.close()
+      await fs.remove(tmpFilename)
```

Comment:
made this async instead of sync for better windows support

Reply:
What the catch with sync `fs.remove` on windows?